File: backend/retrieval_graph/state.py
"""State management for the retrieval graph.

This module defines the state structures used in the retrieval graph. It includes
definitions for agent state, input state, and router classification schema.
"""

from dataclasses import dataclass, field
from typing import Annotated, Literal

from langchain_core.documents import Document
from langchain_core.messages import AnyMessage
from langgraph.graph import add_messages
from typing_extensions import TypedDict

from backend.utils import reduce_docs


# Optional, the InputState is a restricted version of the State that is used to
# define a narrower interface to the outside world vs. what is maintained
# internally.
@dataclass(kw_only=True)
class InputState:
    """Represents the input state for the agent.

    This class defines the structure of the input state, which includes
    the messages exchanged between the user and the agent. It serves as
    a restricted version of the full State, providing a narrower interface
    to the outside world compared to what is maintained internally.
    """

    messages: Annotated[list[AnyMessage], add_messages]
    """Messages track the primary execution state of the agent.

    Typically accumulates a pattern of Human/AI/Human/AI messages; if
    you were to combine this template with a tool-calling ReAct agent pattern,
    it may look like this:

    1. HumanMessage - user input
    2. AIMessage with .tool_calls - agent picking tool(s) to use to collect
         information
    3. ToolMessage(s) - the responses (or errors) from the executed tools
    
        (... repeat steps 2 and 3 as needed ...)
    4. AIMessage without .tool_calls - agent responding in unstructured
        format to the user.

    5. HumanMessage - user responds with the next conversational turn.

        (... repeat steps 2-5 as needed ... )
    
    Merges two lists of messages, updating existing messages by ID.

    By default, this ensures the state is "append-only", unless the
    new message has the same ID as an existing message.

    Returns:
        A new list of messages with the messages from `right` merged into `left`.
        If a message in `right` has the same ID as a message in `left`, the
        message from `right` will replace the message from `left`."""


class Router(TypedDict):
    """Classify user query."""

    logic: str
    type: Literal["more-info", "langchain", "general"]


# This is the primary state of your agent, where you can store any information


@dataclass(kw_only=True)
class AgentState(InputState):
    """State of the retrieval graph / agent."""

    router: Router = field(default_factory=lambda: Router(type="general", logic=""))
    """The router's classification of the user's query."""
    steps: list[str] = field(default_factory=list)
    """A list of steps in the research plan."""
    documents: Annotated[list[Document], reduce_docs] = field(default_factory=list)
    """Populated by the retriever. This is a list of documents that the agent can reference."""
    answer: str = field(default="")
    """Final answer. Useful for evaluations"""
    query: str = field(default="")

File: backend-js/src/retrieval_graph/state.ts
/**
 * State management for the retrieval graph.
 *
 * This module defines the state structures used in the retrieval graph. It includes
 * definitions for agent state, input state, and reducer functions.
 */

import { BaseMessage } from '@langchain/core/messages'
import { Document } from '@langchain/core/documents'
import { Annotation, messagesStateReducer } from '@langchain/langgraph'
import { reduceDocs } from '../utils.js'

/**
 * Shared channel definitions to ensure all annotations use the same channel instances.
 * This prevents "Channel already exists with a different type" errors.
 */
const messagesChannel = Annotation<BaseMessage[]>({
  reducer: messagesStateReducer,
  default: () => [],
})

const documentsChannel = Annotation<Document[], Document[] | 'delete'>({
  reducer: reduceDocs,
  default: () => [],
})

const answerChannel = Annotation<string>({
  reducer: (existing, update) => {
    // Always return a string value for LangSmith compatibility
    if (update !== undefined && update !== null) {
      return String(update)
    }
    if (existing !== undefined && existing !== null) {
      return String(existing)
    }
    return ''
  },
  default: () => '',
})

const queryChannel = Annotation<string>({
  reducer: (existing, update) => {
    // Always return a string value for LangSmith compatibility
    if (update !== undefined && update !== null) {
      return String(update)
    }
    if (existing !== undefined && existing !== null) {
      return String(existing)
    }
    return ''
  },
  default: () => '',
})

/**
 * InputState represents the input to the agent.
 *
 * This is a restricted version of the State that is used to define
 * a narrower interface to the outside world vs. what is maintained internally.
 */
export const InputStateAnnotation = Annotation.Root({
  /**
   * Messages track the primary execution state of the agent.
   *
   * Typically accumulates a pattern of Human/AI/Human/AI messages.
   * Uses messagesStateReducer to merge messages by ID.
   */
  messages: messagesChannel,
})

/**
 * AgentState is the primary state of the retrieval agent.
 *
 * It extends InputState with additional internal state for research planning
 * and document retrieval, matching Python's class AgentState(InputState) pattern.
 */
export const AgentStateAnnotation = Annotation.Root({
  /**
   * Inherit messages from InputStateAnnotation
   */
  ...InputStateAnnotation.spec,

  /**
   * A list of steps in the research plan.
   */
  steps: Annotation<string[]>({
    reducer: (existing, update) => {
      if (update === null || update === undefined) {
        return existing || []
      }
      return update
    },
    default: () => [],
  }),

  /**
   * Documents retrieved by the researcher.
   */
  documents: documentsChannel,

  /**
   * Final answer. Useful for evaluations.
   */
  answer: answerChannel,

  /**
   * The original query from the user.
   */
  query: queryChannel,
})

// Type exports for use in other modules
export type InputState = typeof InputStateAnnotation.State
export type AgentState = typeof AgentStateAnnotation.State
